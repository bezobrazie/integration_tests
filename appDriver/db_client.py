from psycopg2 import Error


class DBClient:
    """
    Клиент для рабоыт с БД.
    connection описан в staticmethod connection
    """

    def __init__(self, connection):
        self.connection = connection

    def get_users(self) -> list:
        """
        Получение всех пользователей
        :return: возвращает список всех пользователей в базе
        """
        with self.connection.cursor() as cursor:
            cursor.execute("SELECT * from users")
            records = cursor.fetchall()

            users = list()
            for row in records:
                users.append(
                    {
                        "id": row[0],
                        "first_name": row[2],
                        "last_name": row[3],
                        "patronymic": row[4],
                        "email": row[5],
                        "phone_number": row[6]
                    }
                )

            return users

    def delete_user(self, email: str):
        """
        Удаление пользователя по email
        :param email: почта пользователя которого хотите удалить из базы
        """
        try:
            with self.connection.cursor() as cursor:
                cursor.execute(f"delete from users where email = '{email}'")
                cursor.execute(f"delete from \"AspNetUsers\" where user_name = '{email}'")
                self.connection.commit()
                count = cursor.rowcount
                print(count, "Запись успешно удалена")
        except (Exception, Error) as error:
            print("Ошибка при работе с PostgreSQL", error)
