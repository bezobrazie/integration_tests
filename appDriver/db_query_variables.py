from typing import Callable

from psycopg2.sql import SQL


class DBQueryVariables:
    """
    Класс для хранения переменных в которых содержаться заготовленные запросы.
    """

    # запрос для выбора всех пользователей
    SELECT_ALL_USERS_QUERY = SQL("SELECT * FROM users;")

    # лямбда функция на получение запроса для удаления пользователя по email
    DELETE_USER_BY_EMAIL_QUERY: Callable[[str], str] = lambda email: f"delete from \"AspNetUsers\" where user_name = '{email}'"
