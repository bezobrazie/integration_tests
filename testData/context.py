import json
import codecs
from copy import deepcopy
from typing import Any


class TestContext:
    """Контекст тестового сценария, копирует данные для переиспользования его в тестах."""

    def __init__(self):
        self.__context_data = dict()

    def add(self, key: str, value: Any) -> None:
        """
        Добавить данные в контекст
            :param key: ключ под которым значение будет сохранено в контекст
            :param value: любой объект который будет добавлен в значение под ключем который указан первым аргументом
        """
        self.__context_data[key] = value

    def get(self, key: str) -> Any:
        """
        Получить данные из контекста
            :param key: ключ по которому будет получен объект из контекста
            :return: запрашиваемые данные из контекста
        """
        return self.__context_data.get(key)

    def get_dict(self):
        """Получить копию словаря с данными контекста"""
        return deepcopy(self.__context_data)

    @staticmethod
    def from_dict(data: dict) -> Any:
        """
        Создать контекст с данными из словаря
            :param data: словарь с данными
            :return: экземпляр контекста с данными из словаря
        """
        context = TestContext()
        for key, value in data.items():
            context.add(key, value)
        return context

    @staticmethod
    def from_json_file(path_to_file: str) -> Any:
        """
        Создать контекст с данными из json файла
            :param path_to_file: путь до json файла
            :return: экземпляр контекста с данными из json файла
        """
        with codecs.open(path_to_file, 'r', encoding='utf-8') as json_file:
            json_content: dict = json.load(json_file)
        return TestContext.from_dict(json_content)
