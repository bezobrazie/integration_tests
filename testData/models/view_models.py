from typing import Optional
from pydantic import BaseModel


class RegisterVM(BaseModel):
    """ Модель представления данных для регистрации"""
    email: str
    password: str
    confirmPassword: str
    firstName: str
    lastName: str
    patronymic: Optional[str]
    phoneNumber: str


class LoginVM(BaseModel):
    """ Модель представления данных для авторизации"""
    email: str
    password: str
