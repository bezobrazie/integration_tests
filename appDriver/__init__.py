from appDriver.dbDriver.db_client import DBClient
from appDriver.dbDriver.db_cleaner import DBCleaner
from appDriver.dbDriver.db_seeder import DBSeeder
from appDriver.httpDriver.http_client import HttpClientOWF
