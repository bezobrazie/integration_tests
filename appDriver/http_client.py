import requests
from urllib.parse import urljoin

from testData.models.view_models import RegisterVM, LoginVM
from appDriver.endpoint_variables import EndpointVariables


class HttpClientOWF:
    """
    Класс для работы с HTTP запросами
    base_url - базовый урл на который будут поступать запросы
    http://localhost:8000/swagger/index.html
    http://localhost:8000/api-docs/index.html
    """

    def __init__(self, base_url):
        self.base_url = base_url

    def register(self, register_vm: RegisterVM) -> requests.Response:
        """
        Метод для регистрации пользователя
        :param register_vm: параметр с моделью данных для регистрации
        :return: объект с информацией об ответе на запрос
        """
        return requests.post(url=urljoin(self.base_url, EndpointVariables.REGISTER_ROUTE), json=register_vm.dict())

    def login(self, login_vm: LoginVM) -> requests.Response:
        """
        Метод для авторизации пользователя
        :param login_vm: параметр с моделью данных для авторизации
        :return: объект с информацией об ответе на запрос
        """
        return requests.post(url=urljoin(self.base_url, EndpointVariables.LOGIN_ROUTE), json=login_vm.dict())
