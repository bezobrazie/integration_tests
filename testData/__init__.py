from .context import TestContext
from testData.scenarios.identity_scenarios.identity_data import IdentityData
