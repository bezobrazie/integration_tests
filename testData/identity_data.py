from testData import TestContext
from testData.models.view_models import RegisterVM, LoginVM
from testData.models.parametrize_models import ParametrizeModel


class IdentityData:
    """
    Класс хранения данных для тестирования
    """

    # Данные используемые в тестах по регистрации с некорректным подтверждением пароля.
    INVALID_REGISTRATION_DATA = TestContext.from_dict(
        {
            "bad_confirm_password": RegisterVM.parse_obj({
                "email": "post@mail.ru",
                "password": "Test123456",
                "confirmPassword": "Test1234567",
                "firstName": "Борис",
                "lastName": "Ельцын",
                "patronymic": "Тестович",
                "phoneNumber": "+79657020827"}),
            "bad_email": RegisterVM.parse_obj({
                "email": "postmail.ru",
                "password": "Test123456",
                "confirmPassword": "Test123456",
                "firstName": "Борис",
                "lastName": "Ельцын",
                "patronymic": "Тестович",
                "phoneNumber": "+79657020827"}),
            "bad_phone": RegisterVM.parse_obj({
                "email": "post@mail.ru",
                "password": "Test123456",
                "confirmPassword": "Test123456",
                "firstName": "Борис",
                "lastName": "Ельцын",
                "patronymic": "Тестович",
                "phoneNumber": "+49657020827"})

        })

    # Данные используемые для успешной решистрации
    VALID_REGISTRATION_DATA = TestContext.from_dict(
        {
            'valid_data': RegisterVM.parse_obj({
                "email": "post@mail.ru",
                "password": "Test123456",
                "confirmPassword": "Test123456",
                "firstName": "Борис",
                "lastName": "Ельцын",
                "patronymic": "Тестович",
                "phoneNumber": "+79657020827"}),
            'valid_data_without_patronymic': RegisterVM.parse_obj({
                "email": "post@mail.ru",
                "password": "Test123456",
                "confirmPassword": "Test123456",
                "firstName": "Борис",
                "lastName": "Ельцын",
                "phoneNumber": "+79657020827"})
        })

    # Данные для авторизации валидные/невалидные
    AUTORIZATION_DATA = TestContext.from_dict({
        "valid_data": LoginVM.parse_obj({
            "email": "post@mail.ru",
            "password": "Test123456"}),
        "bad_login": LoginVM.parse_obj({
            "email": "post@mail1.ru",
            "password": "Test123456"}),
        "bad_password": LoginVM.parse_obj({
            "email": "post@mail.ru",
            "password": "Test12345"})
    })

    # Данные используемые в параметризации для негативных тестов регистрации
    DATA_FOR_BAD_REG = [
        TestContext.from_dict({
            "case": ParametrizeModel.parse_obj({
                "input": INVALID_REGISTRATION_DATA.get("bad_confirm_password"),
                "expected": "Пароли не совпадают",
                "case_name": "Registering with mismatched passwords"})}),
        TestContext.from_dict({
            "case": ParametrizeModel.parse_obj({
                "input": INVALID_REGISTRATION_DATA.get("bad_email"),
                "expected": "Электронный адрес должен соответствовать регулярному выражению ^[-\\w.]+@([A-z0-9][-A-z0-9]+\\.)+[A-z]{2,}$",
                "case_name": "Email does not pass regular expression"})}),
        TestContext.from_dict({
            "case": ParametrizeModel.parse_obj({
                "input": INVALID_REGISTRATION_DATA.get("bad_phone"),
                "expected": "Номер телефона должен соответствовать регулярному выражению ^\\+7\\d{10}$",
                "case_name": "Phone number doesn't pass regular expression"})})]

    # Данные используемые в параметризации для негативных тестов авторизации, обернуты в контекст
    DATA_FOR_BAD_AUTH = [
        TestContext.from_dict({
            'case': ParametrizeModel.parse_obj({
                "input": AUTORIZATION_DATA.get("bad_login"),
                "expected": f"Пользователь с email {AUTORIZATION_DATA.get('bad_login').email} не найден",
                "case_name": "Login with incorrect login"})}),
        TestContext.from_dict({
            'case': ParametrizeModel.parse_obj({
                "input": AUTORIZATION_DATA.get("bad_password"),
                "expected": "Неверный пароль",
                "case_name": "Login with incorrect password"})})]
