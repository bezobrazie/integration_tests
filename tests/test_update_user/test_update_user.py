import pytest
import requests
from syrupy.filters import paths

from testData import TestContext
from testData.models.view_models import LoginVM, UpdateUserVM
from appDriver import HttpClientOWF
from testData.scenarios.update_user_scenario import SUCCESS_UPDATE_SCENARIO


@pytest.mark.incremental
@pytest.mark.usefixtures('seed_users_before_scenario',
                         'seed_addresses_before_scenario',
                         'seed_passports_before_scenario',
                         'http_client')
class TestUpdateUser:
    context = TestContext()

    def test_login(self, http_client: HttpClientOWF):
        data_for_login: LoginVM = LoginVM.parse_obj(SUCCESS_UPDATE_SCENARIO.get('data_for_login'))
        response: requests.Response = http_client.login(data_for_login)
        self.context.add('accessToken', response.json()['accessToken'])

    def test_update_user(self, http_client: HttpClientOWF):
        """
        Не работает конвертация даты из json в datatime объект
        """
        data_for_update_user: UpdateUserVM = UpdateUserVM.parse_obj(SUCCESS_UPDATE_SCENARIO.get('data_for_update_user'))
        token = self.context.get('accessToken')
        response: requests.Response = http_client.update_user_requisites(token, data_for_update_user)
        assert response.status_code == 200, "Код ответа в методе обновления юзера не равен 200"

    def test_get_user_requisites(self, http_client: HttpClientOWF, snapshot):
        token = self.context.get('accessToken')
        response: requests.Response = http_client.get_user_requisites(token)
        try:
            content = response.json()
        except:
            assert False, "Реквизиты пользователя не удалось преобразовать в json"
        # paths позволяет игнорировать поля при добавлении в снапшот
        assert content == snapshot(exclude=paths("address.id", "passport.id"))
