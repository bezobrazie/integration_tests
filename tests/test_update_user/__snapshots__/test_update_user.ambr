# name: TestUpdateUser.test_get_user_requisites
  dict({
    'accounts': list([
    ]),
    'address': dict({
      'apartment': 'Квартира',
      'area': 'Ареа',
      'city': 'Калуга',
      'house': 'Дом',
      'housing': 'Хаусинг',
      'locality': 'Локалити',
      'postIndex': 248915,
      'regionCode': 41,
      'regionName': 'Регион Нэйм',
      'street': 'Улица',
    }),
    'email': 'post@mail.ru',
    'firstName': 'Борис',
    'id': 'e42557a0-cfb2-4a1e-a23d-274f4f86ea39',
    'lastName': 'Ельцын',
    'passport': dict({
      'birthDate': '1994-12-16',
      'birthPlace': 'Калуга',
      'departmentCode': '666-777',
      'issueDate': '2011-11-11',
      'issuer': 'ишюер',
      'number': '123321',
      'series': '4444',
    }),
    'patronymic': 'Тестович',
    'phoneNumber': '+70348819691',
    'registrationDate': '2022-02-08T00:00:00Z',
    'vendors': list([
    ]),
  })
# ---
